src/VisualStudio/Core/Def/Implementation/Venus/ContainedDocument.DocumentServiceProvider.cs

```diff
@@ -0,0 +1,85 @@
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Text.Projection;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Venus
+{
+    internal sealed partial class ContainedDocument
+    {
+        public class DocumentServiceProvider : IDocumentServiceProvider
+        {
+            public static readonly IDocumentServiceProvider Instace = new DocumentServiceProvider();
+
+            private readonly SpanMapper _service;
+
+            private DocumentServiceProvider()
+            {
+                _service = new SpanMapper();
+            }
+
+            public TService GetService<TService>() where TService : IDocumentService
+            {
+                if (_service is TService service)
+                {
+                    return service;
+                }
+
+                return default;
+            }
+
+            private class SpanMapper : ISpanMapperService
+            {
+                public async Task<ImmutableArray<DocumentSpan>> MapSpansAsync(Document document, IEnumerable<TextSpan> spans, CancellationToken cancellationToken)
+                {
+                    var containedDocument = ContainedDocument.TryGetContainedDocument(document.Id);
+                    if (containedDocument == null)
+                    {
+                        return spans.Select(s => new DocumentSpan(document, s)).ToImmutableArray();
+                    }
+
+                    var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                    var editorSnapshot = sourceText.FindCorrespondingEditorTextSnapshot() as IProjectionSnapshot;
+                    if (editorSnapshot == null)
+                    {
+                        return spans.Select(s => new DocumentSpan(document, s)).ToImmutableArray();
+                    }
+
+                    var builder = ArrayBuilder<DocumentSpan>.GetInstance();
+                    foreach (var span in spans)
+                    {
+                        foreach (var primarySpan in editorSnapshot.MapToSourceSnapshots(span.ToSpan()))
+                        {
+                            // this is from http://index/?query=MapSecondaryToPrimarySpan&rightProject=Microsoft.VisualStudio.Editor.Implementation&file=VsTextBufferCoordinatorAdapter.cs&line=177
+                            // make sure we only consider one on primary buffer
+                            if (containedDocument.DataBuffer != primarySpan.Snapshot.TextBuffer)
+                            {
+                                continue;
+                            }
+
+                            // and consider span that is not split
+                            if (primarySpan.Length != span.Length)
+                            {
+                                continue;
+                            }
+
+                            builder.Add(new DocumentSpan(document, primarySpan.Span.ToTextSpan()));
```

Comment:
This doesn't seem right: the spans here now don't correspond to the document itself. The document in this case is the secondary buffer. More generally, it seems like this should be returning some pair more like "filename + span", because in the closed file case we're trying to fix, the .cshtml will have no traditional document representation.

Reply:
Think about this this way: DocumentSpan (for perf reasons) doesn't have an assert that the span is within the document's text. If it did...this would be throwing.

Reply:
with Ryan's change, all razor file will have document. and won't have filepath. currently it works since filepath of the document we have is for the cshtml file. that's why it currently works.

can't do other thing since this whole our system (find all references) expects document unfortunately.